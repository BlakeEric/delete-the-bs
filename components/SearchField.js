import React, { Component } from 'react';
import ReactDOM from 'react-dom'


export default class SearchField extends Component {

  state = {
    valid: false,
    alert: null,
    url: '',
    touched: false,
  }

  handleError(err) {
    alert("Oops, Something went wrong! Please try again");
  }

 /*
  * Wrapper function for submitUrl callback prop
  */
  handleSubmit = (event) => {

    event.preventDefault()

    this.setState({
      touched: false
    })

    if (!this.isValid())
      return

    this.props.submitUrl(this.state.url);
  }

 /*
  * Keep track of controlled form in state
  */
  setUrl = (urlInput) => {
    this.setState({
      url: urlInput,
      touched: true
    })
  }

 /*
  * Validate the url field
  */
  isValid = () => {
    const pattern = /^(?:(?:https?|HTTPS?):\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-zA-Z\u00a1-\uffff0-9]-*)*[a-zA-Z\u00a1-\uffff0-9]+)(?:\.(?:[a-zA-Z\u00a1-\uffff0-9]-*)*[a-zA-Z\u00a1-\uffff0-9]+)*(?:\.(?:[a-zA-Z\u00a1-\uffff]{2,}))\.?)(?::\d{2,5})?(?:[/?#]\S*)?$/
    return pattern.test(this.state.url);
  }

  render() {
    return (
      <div id="searchWrapper">
        <form onSubmit={this.handleSubmit}>
          <input
            type="text"
            onChange={(event) => this.setUrl(event.target.value)}
            placeholder="Paste a website URL to delete bullshit"
            value={this.state.url}
          />
          <button type="submit"
            className="contactForm-submitButton"
            disabled={!this.state.touched || this.props.isLoading || !this.isValid()}
          >
            {this.props.isLoading ? 'â€¦' : 'Submit'}
          </button>
        </form>
      </div>
    )
  }

}

SearchField.defaultProps = {
  isLoading: false
}
