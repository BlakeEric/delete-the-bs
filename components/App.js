import React, { Component } from 'react';
import SearchField from './SearchField'
import Content from './Content'
import getConfig from 'next/config'

const { publicRuntimeConfig = {} } = getConfig() || {};
const baseUrl = publicRuntimeConfig.baseUrl || process.env.BASE_URL;



export default class App extends Component {

 /*
  * Return an object of blank state values.
  * remountKey will be unique each time the object is regenerated
  * and will force remound of SearchField
  */
  initialState = () => {
    return {
      content: null, //object with disc, image, text, copyright
      isLoading: false,
      error: null,
      remountKey: (new Date()).getTime(),
    }
  }

  state = this.initialState();

 /*
  * Make an API request to scrape the url webpage content
  */
  submitUrl = (url) => {

    this.setState({
      isLoading: true,
      error: null,
      content: ''
    });

    fetch(encodeURI(`${baseUrl}/api/scrape?url=${url}`))
    .then(res => {
      if (!res.ok)
        throw "Request error"

      return res.json()

    })
    .then((data) => {
      this.setState({content: data})
    })
    .then(() => {
      this.setState({
        isLoading: false,
      })
    })
    .catch((err) => {
      this.setState({
        error: err,
        isLoading: false
      });
    })
  }

  /*
   * Reset App to it's initial state
   */
  reset = () => {
    this.setState(this.initialState());
  }


  render() {
    return (
      <>
        <section className={`pageInner-header ${this.state.isLoading || this.state.content || this.state.error ? 'top' : ''}`}>
          <div className="pageInner-titleWrap">
            <h1 className="title">Delete the Bullshit</h1>
            <h3 className="subTitle">
              View the content of any website distraction-free
              (without ads, trackers, or other bullshit).
            </h3>
          </div>
          <SearchField key={this.state.remountKey} submitUrl={this.submitUrl} isLoading={this.state.isLoading} />
        </section>

        {this.state.content && !this.state.isLoading ?
          <Content content={this.state.content} isLoading={this.state.isLoading} url={this.state.url} reset={this.reset}/>
        : ''}

        {this.state.isLoading ?
          <div className="loadingText">
            <h2>Deleting Bullshit</h2>
          </div>
        : ''}

        {this.state.error ?
          <div className="errorMessage">
            <h2>Oops, that didn't work...</h2>
            <p>Reason: {this.state.error}</p>
            <p>Make sure you submitted a valid url or try a different one</p>
          </div>
        : ''}
      </>
    )
  }
}
