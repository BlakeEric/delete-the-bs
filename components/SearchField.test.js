/* eslint-env jest */
import { shallow, mount } from 'enzyme'
import ReactDOM from 'react-DOM'
import toJson from 'enzyme-to-json'

import SearchField from './SearchField'
import App from './App'

const mockProps = {
  url: {
    value: "",
    touched: false
  },
  content: null,
  isLoading: false,
  error: null,
  setUrl: () => {}
}

describe('Search Field', () => {

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<SearchField {...mockProps} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('tracks input value in state', () => {
    // The app component holds url value in state
    const wrapper = mount(<App />)
    const input = wrapper.find('input');
    input.instance().value = "field input string";
    input.simulate('change');
    expect(wrapper.state("url").value).toEqual("field input string")
  })

  it('renders a disabled submit button on mount', () => {
    const wrapper = mount(<SearchField {...mockProps} />)
    expect(toJson(wrapper)).toMatchSnapshot();
  })

  it('enables the submit button if a valid url is supplied', () => {
    // The app component holds url value in state
    const wrapper = mount(<App />)
    const input = wrapper.find('input');
    input.instance().value = "https://validurl.com";
    input.simulate('change');

    expect(toJson(wrapper)).toMatchSnapshot();
  })

  it('disables the submit button if an invalid url is supplied', () => {
    // The app component holds url value in state
    const wrapper = mount(<App />)
    const input = wrapper.find('input');
    input.instance().value = "brokenurl.com";
    input.simulate('change');

    expect(toJson(wrapper)).toMatchSnapshot();
  })

})
