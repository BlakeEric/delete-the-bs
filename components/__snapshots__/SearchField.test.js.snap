// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Field disables the submit button if an invalid url is supplied 1`] = `
<SearchField
  isLoading={false}
>
  <div
    id="searchWrapper"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        placeholder="Paste a website URL to delete bullshit"
        type="text"
        value="brokenurl.com"
      />
      <button
        className="contactForm-submitButton"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</SearchField>
`;

exports[`Search Field enables the submit button if a valid url is supplied 1`] = `
<SearchField
  isLoading={false}
>
  <div
    id="searchWrapper"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        placeholder="Paste a website URL to delete bullshit"
        type="text"
        value="https://validurl.com"
      />
      <button
        className="contactForm-submitButton"
        disabled={false}
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</SearchField>
`;

exports[`Search Field renders a disabled submit button on mount 1`] = `
<SearchField
  isLoading={false}
>
  <div
    id="searchWrapper"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        placeholder="Paste a website URL to delete bullshit"
        type="text"
        value=""
      />
      <button
        className="contactForm-submitButton"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</SearchField>
`;
